File: Corrected Keyboard Schmeeboard3/ViewController.m
//
//  ViewController.m
//  Corrected Keyboard Schmeeboard2 http://github.com/apontious/Keyboard-Schmeeboard/
//
//  Created by Andrew Pontious on 1/5/17.
//  Copyright (c) 2017 Andrew Pontious.
//  Some right reserved: http://opensource.org/licenses/mit-license.php
//

#import "ViewController.h"

// http://stackoverflow.com/questions/32125653/ios-foreign-language-keyboard-heights

@interface ViewController ()

@property (nonatomic) IBOutlet UIView *grayView;
@property (nonatomic) IBOutlet UITextField *textField;

@property (nonatomic) IBOutlet NSLayoutConstraint *bottomLayoutConstraint;

@property (nonatomic) BOOL inPhoneRotation;

@end

@implementation ViewController

// Convenience method to turn keyboard notification into usable parameters.
- (void)handleKeyboardNotification:(NSNotification *)notification handler:(void (^)(const CGRect frameBegin, const CGRect frameEnd, const double animationDuration, const UIViewAnimationOptions animationOptions))handler {
    if (handler != nil) {
        CGRect temp = [[[notification userInfo] objectForKey:UIKeyboardFrameBeginUserInfoKey] CGRectValue];
        const CGRect frameBegin = [self.view convertRect:temp fromView:nil];
        temp = [[[notification userInfo] objectForKey:UIKeyboardFrameEndUserInfoKey] CGRectValue];
        const CGRect frameEnd = [self.view convertRect:temp fromView:nil];
        const double animationDuration = [[[notification userInfo] objectForKey:UIKeyboardAnimationDurationUserInfoKey] doubleValue];
        const UIViewAnimationCurve animationCurve = [[[notification userInfo] objectForKey:UIKeyboardAnimationDurationUserInfoKey] unsignedIntegerValue];
        
        UIViewAnimationOptions animationOptions = 0;

        switch (animationCurve) {
            case UIViewAnimationCurveEaseInOut:
                animationOptions = UIViewAnimationOptionCurveEaseInOut;
                break;
            case UIViewAnimationCurveEaseIn:
                animationOptions = UIViewAnimationOptionCurveEaseIn;
                break;
            case UIViewAnimationCurveEaseOut:
                animationOptions = UIViewAnimationOptionCurveEaseOut;
                break;
            case UIViewAnimationCurveLinear:
                animationOptions = UIViewAnimationOptionCurveLinear;
                break;
            default:
                break;
        }
        

        handler(frameBegin, frameEnd, animationDuration, animationOptions);
    }
}

-(void)handleTap {
    [self.textField resignFirstResponder];
}

- (void)viewWillTransitionToSize:(CGSize)size withTransitionCoordinator:(id<UIViewControllerTransitionCoordinator>)coordinator {
	[super viewWillTransitionToSize:size withTransitionCoordinator:coordinator];
	if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPhone &&
		self.view.bounds.size.height == size.width && self.view.bounds.size.width == size.height) {
		self.inPhoneRotation = YES;
	}
	[coordinator animateAlongsideTransition:^(id<UIViewControllerTransitionCoordinatorContext>  _Nonnull context) {
	} completion:^(id<UIViewControllerTransitionCoordinatorContext>  _Nonnull context) {
		self.inPhoneRotation = NO;
	}];
}

- (void)willShowKeyboard:(NSNotification *)notification {
    [self handleKeyboardNotification:notification handler:^(const CGRect frameBegin, const CGRect frameEnd, const double animationDuration, const UIViewAnimationOptions animationOptions) {
        UIViewAnimationOptions modifiedAnimationOptions = animationOptions | UIViewAnimationOptionBeginFromCurrentState;
		// http://stackoverflow.com/questions/12622424/how-do-i-animate-constraint-changes
		[self.view layoutIfNeeded];
		
        [UIView animateWithDuration:animationDuration delay:0 options:modifiedAnimationOptions animations:^(){
			self.bottomLayoutConstraint.constant = CGRectGetHeight(self.view.bounds) - CGRectGetMinY(frameEnd);
			[self.view layoutIfNeeded];
        } completion:nil];
    }];
}
- (void)willHideKeyboard:(NSNotification *)notification {
    [self handleKeyboardNotification:notification handler:^(const CGRect frameBegin, const CGRect frameEnd, const double animationDuration, const UIViewAnimationOptions animationOptions) {
		if (self.inPhoneRotation) {
			return;
		}
		[self.view layoutIfNeeded];
		
        [UIView animateWithDuration:animationDuration delay:0 options:animationOptions animations:^{
			self.bottomLayoutConstraint.constant = CGRectGetHeight(self.view.bounds) - CGRectGetMinY(frameEnd);
			[self.view layoutIfNeeded];
        } completion:nil];
    }];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    UITapGestureRecognizer *tapGestureRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(handleTap)];
    [self.view addGestureRecognizer:tapGestureRecognizer];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];

    NSNotificationCenter *center = [NSNotificationCenter defaultCenter];
    [center addObserver:self selector:@selector(willShowKeyboard:) name:UIKeyboardWillShowNotification object:nil];
    [center addObserver:self selector:@selector(willHideKeyboard:) name:UIKeyboardWillHideNotification object:nil];
}

- (void)viewDidDisappear:(BOOL)animated {
    [super viewDidDisappear:animated];
    
    NSNotificationCenter *center = [NSNotificationCenter defaultCenter];
    [center removeObserver:self name:UIKeyboardWillShowNotification object:nil];
    [center removeObserver:self name:UIKeyboardWillHideNotification object:nil];
}

- (UIInterfaceOrientationMask)supportedInterfaceOrientations {
    return UIInterfaceOrientationMaskAll;
}

@end

File: Corrected Keyboard Schmeeboard4/ViewController.swift
//
//  ViewController.swift
//  Corrected Keyboard Schmeeboard4 http://github.com/apontious/Keyboard-Schmeeboard/
//
//  Created by Andrew Pontious on 1/5/17.
//  Copyright © 2017 Andrew Pontious.
//  Some right reserved: http://opensource.org/licenses/mit-license.php
//

import UIKit

// http://stackoverflow.com/questions/32125653/ios-foreign-language-keyboard-heights

typealias FrameBeginRect = CGRect
typealias FrameEndRect = CGRect
typealias AnimationDuration = Double

class ViewController: UIViewController {
	
	@IBOutlet weak var grayView: UIView!
	@IBOutlet weak var textField: UITextField!
	@IBOutlet weak var bottomLayoutConstraint: NSLayoutConstraint!
	
	private var inRotation: Bool = false
	
	// Convenience method to turn keyboard notification into usable parameters.
	private func handleKeyboardNotification(_ notification: NSNotification,
	                                        handler: (FrameBeginRect, FrameEndRect, AnimationDuration, UIViewAnimationOptions) -> Void) {
		
		// Too giant of a guard statement?
		// I couldn't find a way to combine the check-for-type line and the assign line.
		// I tried a "where" clause (which might not do what I want anyway), but compiler complained.
		// Compiler requires me to use "as!" a.k.a. forced conversion, so I want to test that it's valid beforehand.
		
		guard let userInfo = notification.userInfo,
			// UIKeyboardFrameBeginUserInfoKey
			userInfo[UIKeyboardFrameBeginUserInfoKey] is NSValue?,
			let frameBegin = (userInfo[UIKeyboardFrameBeginUserInfoKey] as! NSValue?)?.cgRectValue,
			// UIKeyboardFrameEndUserInfoKey
			userInfo[UIKeyboardFrameEndUserInfoKey] is NSValue?,
			let frameEnd = (userInfo[UIKeyboardFrameEndUserInfoKey] as! NSValue?)?.cgRectValue,
			// UIKeyboardAnimationDurationUserInfoKey
			userInfo[UIKeyboardAnimationDurationUserInfoKey] is NSNumber?,
			let animationDuration = (userInfo[UIKeyboardAnimationDurationUserInfoKey] as! NSNumber?)?.doubleValue,
			// UIKeyboardAnimationDurationUserInfoKey
			userInfo[UIKeyboardAnimationDurationUserInfoKey] is NSNumber?,
			let animationCurveInteger = (userInfo[UIKeyboardAnimationDurationUserInfoKey] as! NSNumber?)?.intValue,
			let animationCurve = UIViewAnimationCurve(rawValue: animationCurveInteger)
			else {
				return
		}
		
		var animationOptions: UIViewAnimationOptions
		
		switch(animationCurve) {
		case .easeInOut:
			animationOptions = [.curveEaseInOut]
		case .easeIn:
			animationOptions = [.curveEaseIn]
		case .easeOut:
			animationOptions = [.curveEaseOut]
		case .linear:
			animationOptions = [.curveLinear]
		}
		
		handler(frameBegin, frameEnd, animationDuration, animationOptions)
	}
	
	func handleTap() {
		textField.resignFirstResponder()
	}
	
	override func viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator) {
		super.viewWillTransition(to: size, with: coordinator)
		
		if view.bounds.size.height == size.width && view.bounds.size.width == size.height {
			inRotation = true
		}
		coordinator.animate(alongsideTransition: { (context: UIViewControllerTransitionCoordinatorContext) in
		}) { (context: UIViewControllerTransitionCoordinatorContext) in
			self.inRotation = false
		}
	}
	
	func willShowKeyboard(_ notification: NSNotification) {
		handleKeyboardNotification(notification) { (frameBegin: CGRect, frameEnd: CGRect, animationDuration: Double, animationOptions: UIViewAnimationOptions) in
			
			let modifiedAnimationOptions = animationOptions.union([.beginFromCurrentState]);
			// http://stackoverflow.com/questions/12622424/how-do-i-animate-constraint-changes
			view.layoutIfNeeded()
			
			UIView.animate(withDuration: animationDuration, delay: 0, options: modifiedAnimationOptions, animations: {
				self.bottomLayoutConstraint.constant = self.view.bounds.height - frameEnd.minY
				self.view.layoutIfNeeded()
			}, completion: nil)
		}
	}
	
	func willHideKeyboard(_ notification: NSNotification) {
		handleKeyboardNotification(notification) { (frameBegin: CGRect, frameEnd: CGRect, animationDuration: Double, animationOptions: UIViewAnimationOptions) in
			if self.inRotation {
				return
			}
			self.view.layoutIfNeeded()
			
			UIView.animate(withDuration: animationDuration, delay: 0, options: animationOptions, animations: {
				self.bottomLayoutConstraint.constant = self.view.bounds.height - frameEnd.minY
				self.view.layoutIfNeeded()
			}, completion: nil)
		}
	}
	
	override func viewDidLoad() {
		super.viewDidLoad()
		
		let tapGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(ViewController.handleTap))
		self.view.addGestureRecognizer(tapGestureRecognizer)
	}
	
	override func viewDidAppear(_ animated: Bool) {
		super.viewDidAppear(animated)
		
		let center = NotificationCenter.default;
		center.addObserver(self, selector: #selector(ViewController.willShowKeyboard(_:)), name: NSNotification.Name.UIKeyboardWillShow, object: nil)
		center.addObserver(self, selector: #selector(ViewController.willHideKeyboard(_:)), name: NSNotification.Name.UIKeyboardWillHide, object: nil)
	}
	
	override func viewDidDisappear(_ animated: Bool) {
		super.viewDidDisappear(animated)
		
		let center = NotificationCenter.default;
		center.removeObserver(self, name: NSNotification.Name.UIKeyboardWillShow, object: nil)
		center.removeObserver(self, name: NSNotification.Name.UIKeyboardWillHide, object: nil)
	}
	
	override open var supportedInterfaceOrientations: UIInterfaceOrientationMask {
		get {
			return .all
		}
	}
}
